### ISSUE TITLE
Do not gatekeep @context

### ISSUE THREAD

**Issue by puckipedia** [tootsuite/mastodon#8139]
currently, the `@context` requires having an https://www.w3.org/ns/activitystreams value. I do not see the use in this, as I have a custom context that contains the same data, but it won't be accepted by Mastodon, though it's totally valid.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by Gargron** [comment]
No, I am pretty sure that if you have the "same" properties but within a different context they are not the same properties and it's not ActivityPub. cc @cwebber

**Comment by BigBlueHat** [comment]
FWIW, it's possible in JSON-LD (though I don't know about Mastodon's additional requirements) to extend a vocabulary this way:

```json
{
  "@context": [
    "https://www.w3.org/ns/activitystreams",
    {
      "superCustom": "http://example.com/superCustom"
    }
  ],
  "name": "Me",
  "superCustom": 5
}
```
for inline term additions or...
```json
{
  "@context": [
    "https://www.w3.org/ns/activitystreams",
    "http://example.com/superCustomContext"
  ],
  "name": "Me",
  "superCustom": 5
}
```
for an additional remote context file.

There are examples in the tests that do this, so it would seem Mastadon's code is OK with that also (though I'm not sure how it treats the additional terms).

**Comment by nightpool** [comment]
@gargron it's a different context document that uses the same namespace. puckipedias is correct here wrt json-ld, hubzilla (for example) is not, because all of their servers use different namespaces. 

I'm prettttty sure activitystreams doesn't forbid this but I have to check their specific language around extension

**Comment by puckipedia** [comment]
@BigBlueHat I am aware, but I am kinda pedantic and would like my objects to have just `"@context": "https://puckipedia.com/-/context" :P

alternatively, allowing arbitrary `@context` also allows one to use https://www.w3.org/ns/activitystreams-history/, which means the context can be versioned.

the spec says "Implementers SHOULD include the ActivityPub context in their object definitions." -- I consider this close enough imo :P

**Comment by trwnh** [comment]
for proper LD, mastodon should not be checking `@context`, but rather it should be checking what is defined *within* the `@context`. the important bit is that when we say `inbox` we all mean `http://www.w3.org/ns/ldp#inbox`. when we say `content` we all mean `https://www.w3.org/ns/activitystreams#content`. likewise for extensions, when we say `discoverable` we should really be checking for `http://joinmastodon.org/ns#discoverable` (instead of expecting everyone to inline the exact same definition, or even worse, checking for the term itself without a namespace...)

mastodon should ideally fetch and cache contexts, as well as pre-package certain contexts with the software if they are known to be final and unchanging (security v1, for example)

**Comment by BigBlueHat** [comment]
> @BigBlueHat I am aware, but I am kinda pedantic and would like my objects to have just `"@context": "https://puckipedia.com/-/context" :P

Sorry I didn't reply to this...five years ago. 😕 

Basically, doing that prevents the economies of scale provided by cache-friendly, community published (as in made public/shared/aggreed upon) context files. JSON-LD processors won't care and will dereference (or pull from cache) whatever they're given, but each site re-declaring the same values in a different context file means you remove the cache benefits from the tooling used prior to JSON-LD processing.

There's "shared library" value/optimizations you'll miss out on...or cause the ecosystem to miss out on.

**Comment by trwnh** [comment]
> There's "shared library" value/optimizations you'll miss out on...or cause the ecosystem to miss out on.

i'm pretty sure the cache optimization is only really relevant for long-lived and unchanging context definitions. as previously mentioned, security v1 and identity v1 could be cached forever. something like activitystreams could be cached... for some years, at the current rate of change, but could reasonably be refreshed every few months, or possibly just package the latest context document with each release of mastodon. basically, treat context documents like any other dependency.

i'm also pretty sure that the expected way to deal with differing contexts is to first expand according to *their* context so you get the unambiguous representation... and then compact/flatten according to *your* context, so you get something shorthand that you can work with for your application that you understand. anything not understood will remain expanded.

example: https://json-ld.org/playground/#startTab=tab-compacted&json-ld=%5B%7B%22%40id%22%3A%22http%3A%2F%2Fdbpedia.org%2Fresource%2FJohn_Lennon%22%2C%22http%3A%2F%2Fschema.org%2FbirthDate%22%3A%5B%7B%22%40type%22%3A%22http%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema%23date%22%2C%22%40value%22%3A%221940-10-09%22%7D%5D%2C%22http%3A%2F%2Fxmlns.com%2Ffoaf%2F0.1%2Fname%22%3A%5B%7B%22%40value%22%3A%22John%20Lennon%22%7D%5D%2C%22http%3A%2F%2Fschema.org%2Fspouse%22%3A%5B%7B%22%40id%22%3A%22http%3A%2F%2Fdbpedia.org%2Fresource%2FCynthia_Lennon%22%7D%5D%2C%22https%3A%2F%2Fwww.w3.org%2Fns%2Factivitystreams%23name%22%3A%5B%7B%22%40value%22%3A%22John%20Lennon%22%7D%5D%7D%5D&context=%7B%22%40context%22%3A%22https%3A%2F%2Fwww.w3.org%2Fns%2Factivitystreams%22%7D

starting with someone else's context:

```
{
  "@context": "https://json-ld.org/contexts/person.jsonld",
  "@id": "http://dbpedia.org/resource/John_Lennon",
  "name": "John Lennon",
  "born": "1940-10-09",
  "spouse": "http://dbpedia.org/resource/Cynthia_Lennon"
}
```

we expand to get something unambiguous:

```
[
  {
    "@id": "http://dbpedia.org/resource/John_Lennon",
    "http://schema.org/birthDate": [
      {
        "@type": "http://www.w3.org/2001/XMLSchema#date",
        "@value": "1940-10-09"
      }
    ],
    "http://xmlns.com/foaf/0.1/name": [
      {
        "@value": "John Lennon"
      }
    ],
    "http://schema.org/spouse": [
      {
        "@id": "http://dbpedia.org/resource/Cynthia_Lennon"
      }
    ]
  }
]
```

which compacts against activitystreams like so:

```
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "id": "http://dbpedia.org/resource/John_Lennon",
  "http://schema.org/birthDate": {
    "type": "xsd:date",
    "@value": "1940-10-09"
  },
  "http://schema.org/spouse": {
    "id": "http://dbpedia.org/resource/Cynthia_Lennon"
  },
  "http://xmlns.com/foaf/0.1/name": "John Lennon"
}
```

we notice that only `@id` has been changed to `id`, and this object has no properties we understand as shorthand terms. if we prior insert the following into the expanded form:

```
{
  "@type": ["https://www.w3.org/ns/activitystreams#Person"],
  "https://www.w3.org/ns/activitystreams#name": [
    {
      "@value": "John Lennon"
    }
  ]
}
```

then when we compact, we get something we understand as shorthand terms: `type` and `name`.

```
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "id": "http://dbpedia.org/resource/John_Lennon",
  "type": "Person",
  "http://schema.org/birthDate": {
    "type": "xsd:date",
    "@value": "1940-10-09"
  },
  "http://schema.org/spouse": {
    "id": "http://dbpedia.org/resource/Cynthia_Lennon"
  },
  "http://xmlns.com/foaf/0.1/name": "John Lennon",
  "name": "John Lennon"
}
```

as far as our application logic is concerned, this is equivalent:

```
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "id": "http://dbpedia.org/resource/John_Lennon",
  "type": "Person",
  "name": "John Lennon"
}
```

because we discard or ignore any extension we don't understand. but there's no reason to throw away a perfectly good object that you know how to process, just because its context doesn't include some "magic string". in theory, a JSON-only implementation of activitystreams might neglect to include the `@context` entirely, and might rely on `application/activity+json` to convey the same information. an LD-based implementation that wishes to fully support activitystreams needs to be aware of this and inject the context if it is missing, according to AS2-Core 2.1 "JSON-LD":

> When a JSON-LD enabled Activity Streams 2.0 implementation encounters a JSON document identified using the " application/activity+json" MIME media type, and that document does not contain a @context property whose value includes a reference to the normative [Activity Streams 2.0 JSON-LD @context definition](https://www.w3.org/ns/activitystreams), the implementation MUST assume that the normative @context definition still applies.
